File: algorithms/code/leetcode/lc090_subsets_ii/lc090_subsets_ii.py
# https://leetcode.com/problems/subsets-ii/

import unittest
from typing import Any, Callable


class Solution:
    def __subsets(
        self,
        nums: list[int],
        offset: int,
        current: list[int],
        listed: list[list[int]],
    ) -> None:
        listed.append(current.copy())
        i = offset
        while i < len(nums):
            count = 1
            j = i + 1
            while j < len(nums) and nums[j - 1] == nums[j]:
                count += 1
                j += 1
            for _ in range(count):
                current.append(nums[i])
                self.__subsets(nums, i + count, current, listed)
            for _ in range(count):
                current.pop()
            i += count

    def subsetsWithDup(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        listed: list[list[int]] = []
        self.__subsets(nums, 0, [], listed)
        return listed


class TestCode(unittest.TestCase):
    @staticmethod
    def cmp_to_key(mycmp: Callable) -> Any:  # pragma: no cover
        class K:
            def __init__(self, obj: Any, *_: Any) -> None:
                self.obj = obj

            def __lt__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) < 0

            def __gt__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) > 0

            def __eq__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) == 0

            def __le__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) <= 0

            def __ge__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) >= 0

            def __ne__(self, other: Any) -> bool:
                return mycmp(self.obj, other.obj) != 0

        return K

    @staticmethod
    def orderly_comparator(
        l1: list[int], l2: list[int]
    ) -> int:  # pragma: no cover
        difference = len(l1) - len(l2)
        if difference != 0:
            return difference
        for l1s, l2s in zip(l1, l2, strict=True):
            if l1s < l2s:
                return -1
            if l1s > l2s:
                return 1
        return 0

    def __test(
        self, expected: list[list[int]], result: list[list[int]]
    ) -> None:
        result = sorted(
            result, key=TestCode.cmp_to_key(TestCode.orderly_comparator)
        )
        assert len(expected) == len(result)
        for i, expected_i in enumerate(expected):
            assert len(expected_i) == len(result[i])
            result[i].sort()
            for j, expected_i_j in enumerate(expected_i):
                assert expected_i_j == result[i][j]

    def test_1_2_2(self) -> None:
        listed = Solution().subsetsWithDup([1, 2, 2])
        expected = [[], [1], [2], [1, 2], [2, 2], [1, 2, 2]]
        self.__test(expected, listed)

File: algorithms/code/leetcode/lc090_subsets_ii/lc090_subsets_ii.rs
// https://leetcode.com/problems/subsets-ii/
// #medium

pub struct Solution;

impl Solution {
    fn subsets_internal(
        nums: &[i32],
        offset: usize,
        mut current: Vec<i32>,
        list: &mut Vec<Vec<i32>>,
    ) {
        list.push(current.clone());
        let mut i = offset;
        while i < nums.len() {
            let mut count = 1;
            for j in (i + 1)..nums.len() {
                if nums[j - 1] != nums[j] {
                    break;
                }
                count += 1;
            }
            for _ in 0..count {
                current.push(nums[i]);
                Self::subsets_internal(nums, i + count, current.clone(), list);
            }
            for _ in 0..count {
                current.pop();
            }
            i += count;
        }
    }

    pub fn subsets_with_dup(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        nums.sort_unstable();
        let mut list = Vec::new();
        Self::subsets_internal(&nums, 0, Vec::new(), &mut list);
        list
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use std::cmp::Ordering;

    pub fn compare_orderly(l1: &Vec<i32>, l2: &Vec<i32>) -> Ordering {
        let difference = l1.len().cmp(&l2.len());
        if difference != std::cmp::Ordering::Equal {
            return difference;
        }
        for (a, b) in l1.iter().zip(l2.iter()) {
            let difference = a.cmp(b);
            if difference != std::cmp::Ordering::Equal {
                return difference;
            }
        }
        std::cmp::Ordering::Equal
    }

    #[test]
    fn test_122() {
        let expected = vec![
            vec![],
            vec![1],
            vec![2],
            vec![1, 2],
            vec![2, 2],
            vec![1, 2, 2],
        ];
        let mut result = Solution::subsets_with_dup(vec![1, 2, 2]);
        result.sort_by(compare_orderly);
        assert_eq!(expected, result);
    }

    #[test]
    fn test_comparator() {
        assert_eq!(compare_orderly(&vec![1, 2], &vec![1, 2]), Ordering::Equal);
    }
}
